node/consensus/src/transactions_queue.rs

```diff
@@ -0,0 +1,354 @@
+        self.deployments
+            .priority_queue
+            .transactions
+            .clone()
+            .into_iter()
+            .chain(self.deployments.queue.clone())
+            .chain(self.executions.priority_queue.transactions.clone())
+            .chain(self.executions.queue.clone())
+    }
+}
+
+pub struct TransactionsQueueInner<N: Network> {
+    capacity: usize,
+    queue: LruCache<N::TransactionID, Transaction<N>>,
+    priority_queue: PriorityQueue<N>,
+}
+
+impl<N: Network> TransactionsQueueInner<N> {
+    fn new(capacity: usize) -> Self {
+        Self {
+            capacity,
+            queue: LruCache::new(NonZeroUsize::new(capacity).unwrap()),
+            priority_queue: Default::default(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.queue.len().saturating_add(self.priority_queue.len())
+    }
+
+    fn contains(&self, transaction_id: &N::TransactionID) -> bool {
+        self.queue.contains(transaction_id) || self.priority_queue.transactions.contains_key(transaction_id)
+    }
+
+    fn insert(
+        &mut self,
+        transaction_id: N::TransactionID,
+        transaction: Transaction<N>,
+        priority_fee: U64<N>,
+    ) -> Result<()> {
+        // If the queue is not full, insert in the appropriate queue.
+        if self.len() < self.capacity {
+            if priority_fee.is_zero() {
+                self.queue.get_or_insert(transaction_id, || transaction);
+            } else {
+                self.priority_queue.insert(transaction_id, transaction, priority_fee);
+            }
+
+            return Ok(());
+        }
+
+        match (self.priority_queue.len() < self.capacity, *priority_fee) {
+            // Invariant: if the queue is at capacity but the priority queue
+            // isn't equal to the capacity, the low-priority queue must be non-empty.
+            (true, 0) => {
+                let _ = self.queue.get_or_insert(transaction_id, || transaction);
+            }
+            (true, _fee) => {
+                // Remove an entry from the low-priority queue to make room for the high-priority transaction.
+                self.queue.pop_lru();
+                self.priority_queue.insert(transaction_id, transaction, priority_fee)
+            }
+
+            // Invariant: if the queue is at capacity but the priority queue is
+            // equal to the capacity, the low-priority queue must be empty.
+            (false, 0) => bail!("The memory pool is full"),
+            (false, _fee) => self.priority_queue.compare_insert(transaction_id, transaction, priority_fee),
+        }
+
+        Ok(())
+    }
+
+    pub fn pop(&mut self) -> Option<(N::TransactionID, Transaction<N>)> {
+        self.priority_queue.pop().or_else(|| self.queue.pop_lru())
+    }
+}
+
+struct PriorityQueue<N: Network> {
+    /// A counter to ensure fifo ordering for transmissions with the same fee.
+    counter: u64,
```

Comment:
Would using a unix timestamp be easier to manage than a rolling counter?

Reply:
A unix timestamp has a precision of 1 second, that's likely not granular enough to guarantee ordering when there's a lot of inbound traffic?